## 단원별 심화 연습 문제

In [20]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import seaborn as sns
import glob
from IPython.display import Image

# set floating point formatting
pd.options.display.float_format = '{:,.6f}'.format

### Q2

**범위**
- (이전 범위 포함)
- 통계 및 연산

`/mnt/elice/dataset/StudentsPerformance.csv`파일을 읽어서 DataFrame 형식으로 출력합니다. (상위 5개 행을 출력 합니다.)

In [21]:
# 데이터셋 로드
import pandas as pd

students = pd.read_csv('/mnt/elice/dataset/StudentsPerformance.csv')
students.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

다음의 요구사항에 맞는 데이터프레임을 출력하세요

- `students` 데이터프레임의 `race/ethnicity`가 `group B`인 그룹을 필터합니다.
- `gender`가 `male`인 대상을 필터합니다.
- `math score`, `reading score`, `writing score`에 대한 `평균(mean)`과 `표준편차(std)`를 출력합니다.

In [22]:
# 코드를 입력해 주세요
cond1 = students['race/ethnicity'] == 'group B'
cond2 = students['gender'] == 'male'
students.loc[cond1 & cond2, ['math score', 'reading score', 'writing score']].agg(['mean', 'std'])

math score  reading score  writing score
mean   65.930233      62.848837      60.220930
std    14.156928      14.695752      14.854907

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>math score</th>
      <th>reading score</th>
      <th>writing score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>mean</th>
      <td>65.930233</td>
      <td>62.848837</td>
      <td>60.220930</td>
    </tr>
    <tr>
      <th>std</th>
      <td>14.156928</td>
      <td>14.695752</td>
      <td>14.854907</td>
    </tr>
  </tbody>
</table>
</div>

다음의 요구사항에 맞는 데이터프레임을 출력하세요

- `students`데이터프레임에서 `gender`가 `male`인 대상을 필터합니다.
- `lunch`가 `standard` 대상을 필터합니다.
- `reading score`가 80 이상인 데이터를 필터합니다.
- `math score`, `reading score`, `writing score`에 대한 요약통계(describe)를 출력합니다.

In [23]:
# 코드를 입력해 주세요
cond1 = students['gender'] == 'male'
cond2 = students['lunch'] == 'standard'
cond3 = students['reading score'] >= 80
students.loc[cond1 & cond2 & cond3, ['math score', 'reading score', 'writing score']].describe()

math score  reading score  writing score
count   68.000000      68.000000      68.000000
mean    87.897059      85.470588      82.955882
std      6.388281       4.872924       6.391304
min     75.000000      80.000000      71.000000
25%     83.000000      82.000000      78.000000
50%     87.000000      84.000000      82.000000
75%     92.250000      87.000000      86.250000
max    100.000000     100.000000     100.000000

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>math score</th>
      <th>reading score</th>
      <th>writing score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>68.000000</td>
      <td>68.000000</td>
      <td>68.000000</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>87.897059</td>
      <td>85.470588</td>
      <td>82.955882</td>
    </tr>
    <tr>
      <th>std</th>
      <td>6.388281</td>
      <td>4.872924</td>
      <td>6.391304</td>
    </tr>
    <tr>
      <th>min</th>
      <td>75.000000</td>
      <td>80.000000</td>
      <td>71.000000</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>83.000000</td>
      <td>82.000000</td>
      <td>78.000000</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>87.000000</td>
      <td>84.000000</td>
      <td>82.000000</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>92.250000</td>
      <td>87.000000</td>
      <td>86.250000</td>
    </tr>
    <tr>
      <th>max</th>
      <td>100.000000</td>
      <td>100.000000</td>
      <td>100.000000</td>
    </tr>
  </tbody>
</table>
</div>

위의 표를 참고하여 `math score`, `reading score`, `writing score`의 IQR을 산정합니다.
즉, 직전에 요구사항에 맞게 필터링한 데이터프레임을 가지고 아래 내용을 수행하면 됩니다.

- IQR(Inter Quantile Range) = ((3rd Quantile, 75%) - (1st Quantile, 25%)) * 1.5
  * IQR Bottom = (1st Quantile, 25%) - IQR
  * IQR Top: (3rd Quantile, 75%) + IQR
 
 왜 IQR을 구해서 Bottom과 Top을 구하나요? -> Bottom ~ Top 구간 외에 있는 값들을 (잠재적)이상치로 보기 위해서

In [24]:
Image(url='https://img1.daumcdn.net/thumb/R800x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FcAqc6V%2FbtqyQLiddUd%2FiXQVu1nYTo2rx3Q8xZBqy0%2Fimg.png')

위에 명시된 공식을 참고하여 **IQR Bottom**을 산출해 주세요

In [32]:
# 코드를 입력해 주세요
df = students.loc[cond1 & cond2 & cond3, ['math score', 'reading score', 'writing score']].describe()
iqr = (df.loc['75%'] - df.loc['25%']) * 1.5
df.loc['25%'] - iqr

math score      69.125000
reading score   74.500000
writing score   65.625000
dtype: float64

<p><strong>[출력 결과]</strong></p><pre>math score       69.125
reading score    74.500
writing score    65.625
dtype: float64</pre>

위에 명시된 공식을 참고하여 **IQR Top**을 산출해 주세요

In [34]:
# 코드를 입력해 주세요
df.loc['75%'] + iqr

math score      106.125000
reading score    94.500000
writing score    98.625000
dtype: float64

<p><strong>[출력 결과]</strong></p><pre>math score       106.125
reading score     94.500
writing score     98.625
dtype: float64</pre>

## 제출

제출을 위해 새로 로드된 `/mnt/elice/dataset/StudentsPerformance.csv` 데이터셋에서 다음 요구사항에 맞는 데이터프레임을 `result_df`에 저장하세요.

- `students`데이터프레임에서 `gender`가 `male`인 대상을 필터합니다.
- `lunch`가 `standard` 대상을 필터합니다.
- `reading score`가 80 이상인 데이터를 필터합니다.
- `math score`, `reading score`, `writing score`에 대한 요약통계(describe)

In [ ]:
students = pd.read_csv('/mnt/elice/dataset/StudentsPerformance.csv')

cond1 = 

result_df = 

## 채점 수행

아래 코드는 채점 수행을 위한 코드입니다.

따라서 이를 수정했을시 **채점이 제대로 이루어지지 않습니다.**

**주의**: 채점 코드를 실행하기 전에 반드시 코드 파일을 한번 저장하시길 바랍니다.

In [ ]:
import os

result_df.to_json('result.json')

os.system('elice_grade result.json cds_ds_exercise.ipynb')